src/assert.js

```diff
@@ -15,25 +15,38 @@ QUnit.assert = Assert.prototype = {
 		}
 	},
 
-	// Increment this Test's semaphore counter, then return a single-use function that
+	// Increment this Test's semaphore counter, then return a function that
 	// decrements that counter a maximum of once.
-	async: function() {
+	async: function( count ) {
 		var test = this.test,
-			popped = false;
+			popped = false,
+			acceptCallCount = count;
+
+		if ( typeof acceptCallCount === "undefined" ) {
+			acceptCallCount = 1;
+		}
 
 		test.semaphore += 1;
 		test.usedAsync = true;
 		pauseProcessing();
 
 		return function done() {
-			if ( !popped ) {
-				test.semaphore -= 1;
-				popped = true;
-				resumeProcessing();
-			} else {
-				test.pushFailure( "Called the callback returned from `assert.async` more than once",
+			acceptCallCount -= 1;
+
+			if ( acceptCallCount > 0 ) {
+				pauseProcessing();
```

Comment:
@mixed, isn't it already paused before the done function?

https://github.com/jquery/qunit/pull/863/files#diff-f8f3ba9b60ca5c3a0a1dd233f0d34122L26

If `config.testTimeout` is not enough, it should be modified by the user.

---

@mixed I agree that @gibson042's suggestion helps in the function readability. It's like saying to check for fails before running the rest of the code.

In another PoV, the way it is, give a very slight improvement to the performance as popped is checked if the function has not returned with the `pauseProcessing();`

---

If anything changes in the code and for future improvements with the tail call optimization, you can blend the return statements with the last call:

```js
...
return pauseProcessing();

...

return test.pushFailure( ... );

...

return resumeProcessing();
```